File: src/main/ruby/g0201_0300/s0300_longest_increasing_subsequence/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Dynamic_Programming
# #Binary_Search #Algorithm_II_Day_16_Dynamic_Programming #Binary_Search_II_Day_3
# #Dynamic_Programming_I_Day_18 #Udemy_Dynamic_Programming #Big_O_Time_O(n*log_n)_Space_O(n)
# #2023_11_26_Time_68_ms_(96.55%)_Space_211.5_MB_(31.03%)

# @param {Integer[]} nums
# @return {Integer}
def length_of_lis(nums)
  return 0 if nums.nil? || nums.empty?

  dp = Array.new(nums.length + 1, Float::INFINITY)

  left = 1
  right = 1

  nums.each do |curr|
    start_idx = left
    end_idx = right

    # Binary search: find the index where dp[index] is lower than curr
    while start_idx + 1 < end_idx
      mid_idx = start_idx + (end_idx - start_idx) / 2

      if dp[mid_idx] > curr
        end_idx = mid_idx
      else
        start_idx = mid_idx
      end
    end

    # Update the dp table
    if dp[start_idx] > curr
      dp[start_idx] = curr
    elsif curr > dp[start_idx] && curr < dp[end_idx]
      dp[end_idx] = curr
    elsif curr > dp[end_idx]
      dp[end_idx + 1] = curr
      right += 1
    end
  end

  right
end

File: src/main/erlang/g0201_0300/s0300_longest_increasing_subsequence/Solution.erl
% #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Binary_Search
% #Algorithm_II_Day_16_Dynamic_Programming #Binary_Search_II_Day_3 #Dynamic_Programming_I_Day_18
% #Udemy_Dynamic_Programming #Top_Interview_150_1D_DP #Big_O_Time_O(n*log_n)_Space_O(n)
% #2025_01_20_Time_954_ms_(100.00%)_Space_64.54_MB_(100.00%)

-spec length_of_lis(Nums :: [integer()]) -> integer().
length_of_lis(Nums) ->
    Pred = fun(V, Index) -> { {Index, V}, Index + 1} end,
    {IVals, _} = lists:mapfoldl(Pred, 0, Nums),
    [H|T] = IVals,
    Tid = ets:new(lis, []),
    length_of_lis([], T, H, Tid, 0).

length_of_lis(Before, [], {_, Val}, Tid, Max) ->
    max(lis(Before, Val, Tid) + 1, Max);
length_of_lis(Before, [Next|After], Current={Index, Val}, Tid, Max) ->
    Lis = lis(Before, Val, Tid) + 1,
    ets:insert(Tid, {Index, Lis}),
    length_of_lis([Current|Before], After, Next, Tid, max(Max, Lis)).

lis([], _, _) -> 0;
lis(L, N, Tid) ->
    Opts = lists:filter(fun({I, V}) -> V < N end, L),
    Pred = fun({I, _}) -> ets:lookup_element(Tid, I, 2) end,
    case Opts of
        [] -> 0;
        _ ->
            lists:max(lists:map(Pred, Opts))
    end.
